Plug misc finalize leaks on finalize

ggouaillardet opened the pull request.

bosilca commented on ompi/dpm/dpm.c:
```diff
@@ -1057,15 +1057,22 @@ int ompi_dpm_dyn_finalize(void)
         }
 
         if (j != ompi_comm_num_dyncomm+1) {
+            for(i=0; i<j; i++) {
```
I don't see the point of this entire check. If j is less than ompi_comm_num_dyncomm then we should just call disconnect on what we have and continue the normal path. If however j is larger that ompi_comm_num_dyncomm we are already in big troubles as we have overwritten outside the bounds of the temporary objs array.

bosilca replied:
Moreover there should be no opal_output(0,...) in the disconnect_waitall function.

ggouaillardet replied:
I will double check that tomorrow

ggouaillardet commented on opal/mca/pmix/base/pmix_base_select.c:
```diff
@@ -33,11 +36,10 @@ int opal_pmix_base_select(void)
-             "PMIX_MCA_mca_base_component_show_load_errors=%d",
-             mca_base_component_show_load_errors ? 1 : 0);
-    putenv(pmix_show_load_errors_env);
+    char pmix_show_load_errors_env[2];
```
Will do

hjelmn commented on opal/mca/pmix/pmix3x/pmix3x_component.c:
```diff
@@ -97,13 +98,15 @@ static int external_register(void)
                                            OPAL_INFO_LVL_4,
                                            MCA_BASE_VAR_SCOPE_CONSTANT,
                                            &pmix_library_version);
+    free(tmp);
```
This is correct. The MCA base duplicates the string because it becomes owned by the MCA base (to support MPI_T). If the original value was allocated it needs to be freed by the caller.

bosilca replied:
Personal opinion: this is u.g.l.y ;) Why not simply taking ownership on the value and letting the upper level cope with that ?

hjelmn replied:
Because most string variables are constant and so I wanted the ability to set the initial value to a constant string.

ggouaillardet replied:
This is how I read the code :
 - `&pmix_library_version` is saved at line 1495 (note the `&`)
 - Then `pmix_library_version` is replaced with the strdup’ed string at line 1499.

Bottom line, the original `pmix_library_version` is no more referenced, that is why I saved it in `tmp` in order to `free(tmp)`

@hjelmn can you please advise ?

On second thought, a more elegant approach could be to have this handled in MCA via a new flag (no `strdup()` at all or `strdup(); free()`)

bosilca replied:
@hjelmn but then why returning a strdup instead of letting the upper level to make a copy of its own string if/when needed. Instead, we can guarantee that the MCA will not modify and not free the string.

hjelmn replied:
Optimize for the common case and let the uncommon case have the extra code.

hjelmn replied:
This behavior is well-documented.

bosilca replied:
@ggouaillardet is right, nobody's using the original pointer. Moreover, the storage object must be global so that the mbv_storage could point to something that remains valid during the entire execution.

ibm-ompi commented:
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0fe387e7ccd152473285e1647ed62d55

ibm-ompi commented:
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/47728ebfb465f7c2ca362b6076f5443e

ggouaillardet commented:
@bosilca I removed the changes in `ompi/dpm` (I cannot remember what was the rationale for that), and changed the syntax in `pmix_show_load_errors_env` initialization (the syntax you suggested is invalid - at least on my CentOS 7 box with gcc 4.8.5 - so I changed its type to `char *`)

can you please give it a final review before I merge ?

hppritcha commented:
@ggouaillardet  Hmm...should this be merged into v4.0.x?

gpaulsen commented:
May have gone into v4.0.x already.  Howard and I Do want it there.
